schema/scope_test.go

```diff
@@ -498,3 +498,54 @@ func TestApplyingExternalNamespaceToNonRefTypes(t *testing.T) {
 		})
 	}
 }
+
+func TestGetRootObject(t *testing.T) {
+	rootObject := schema.NewObjectSchema("a", map[string]*schema.PropertySchema{})
+	correctSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"a": rootObject,
+		},
+		RootValue: "a",
+	}
+	assert.Equals(t, correctSchema.RootObject(), rootObject)
+}
+
+func TestMismatchedRoot(t *testing.T) {
+	// This is a common user mistake: invalid RootValue in the scope.
+	// Panic when a Scope's RootValue is not a key in its ObjectsValue
+	brokenSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"a": schema.NewObjectSchema("a", map[string]*schema.PropertySchema{}),
+		},
+		RootValue: "wrong",
+	}
+	assert.PanicsContains(t, func() {
+		brokenSchema.RootObject()
+	}, "root object with ID \"wrong\" not found; available objects:\n\ta")
+}
+
+func TestNilRoot(t *testing.T) {
+	// This is just a bug case; nil object in the objects map.
+	brokenSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"a": nil,
+		},
+		RootValue: "a",
+	}
+	assert.PanicsContains(t, func() {
+		brokenSchema.RootObject()
+	}, "root object with ID \"a\" is nil")
+}
+
+func TestMismatchedRootID(t *testing.T) {
+	// This is a common user mistake: valid scope map key doesn't match the object's ID
+	brokenSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"wrong": schema.NewObjectSchema("a", map[string]*schema.PropertySchema{}),
+		},
+		RootValue: "wrong",
+	}
+	assert.PanicsContains(t, func() {
+		brokenSchema.RootObject()
+	}, "root object's ID \"a\" doesn't match its map key \"wrong\"")
+}
```

Comment:
There are three values which are supposed to match (right?):
1. the `schema.ScopeSchema.RootValue`;
2. the `schema.ScopeSchema.ObjectsValue` key; and,
3. that key's value's schema ID.

So, ignoring the case where the key value is `nil`, it seems like there should be three test cases where each of the values is "wrong".  I think that `TestMismatchedRoot()` covers case number 1, and I think that `TestMismatchedRootID()` covers case number 3; however, I think we should have a third test for case number 2, something like:
```go
func TestMismatchedKey(t *testing.T) {
	// This is another possible user mistake: mismatched scope map key
	brokenSchema := schema.ScopeSchema{
		ObjectsValue: map[string]*schema.ObjectSchema{
			"wrong": schema.NewObjectSchema("a", map[string]*schema.PropertySchema{}),
		},
		RootValue: "a",
	}
	assert.PanicsContains(t, func() {
		brokenSchema.RootObject()
	}, "<...>")
}
```

Reply:
Added.